### ISSUE TITLE
Infinite loop in op_dependencies with recurrent layers.

### ISSUE THREAD

**Issue by dekromp** [Thenerdstation/mltest#4]
Hi,
I observed that `op_dependencies` has an infinite loop when applied on a graph that contains a recurrent layer.
Minimal example that reproduces the bug:

```python
import tensorflow as tf


# Simple RNN.
input_ = tf.placeholder(tf.float32, [None, 10, 6], name='input_')
last_hidden_state = tf.keras.layers.SimpleRNN(
    5, activation=tf.nn.tanh)(input_)

# This call causes an infinite loop.
parent_ops = op_dependencies(last_hidden_state)
```

I think this is caused by 
https://github.com/Thenerdstation/mltest/blob/e3e1c97405da660208d6a36be08f85ee1e143698/mltest/mltest.py#L315-L317

Here, you are checking if the op `op_input.op` is in `visited` but the way I understand your code you are filling `visited` with tensors not with ops. 

After changing the `if op_input.op not in visited` to `if op_input not in visited` the loop becomes finite. However I ran into another issue which I could not resolve yet...
Minimal example:

```python
import numpy as np
import tensorflow as tf


input_ = tf.placeholder(tf.float32, [None, 10, 6], name='input_')
last_hidden_state = tf.keras.layers.SimpleRNN(
    5, activation=tf.nn.tanh)(input_)

target = tf.placeholder(tf.float32, [None, 5])
loss = tf.reduce_mean(tf.square(target - last_hidden_state))

# Changed op_dependencies and loop is now finite.
parent_ops = op_dependencies(loss)

feed_dict = {
    input_: np.random.uniform(-1, 1, size=(8, 10, 6)).astype(np.float32),
    target: np.random.uniform(-1, 1, size=(8, 5))}

with tf.Session() as session:
    session.run(tf.global_variables_initializer())
    results = session.run(parent_ops, feed_dict=feed_dict)

```
Which causes the following error:
```
Traceback (most recent call last):
  File "minimal_mltest_bug2.py", line 74, in <module>
    results = session.run(parent_ops, feed_dict=feed_dict)
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 895, in run
    run_metadata_ptr)
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1128, in _run
    feed_dict_tensor, options, run_metadata)
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1344, in _do_run
    options, run_metadata)
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1363, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InternalError: Retval[50] has already been set.
	 [[Node: _retval_simple_rnn/while/TensorArrayWrite/TensorArrayWriteV3/Enter_0_50 = _Retval[T=DT_RESOURCE, index=50, _device="/job:localhost/replica:0/task:0/device:CPU:0"](simple_rnn/while/TensorArrayWrite/TensorArrayWriteV3/Enter)]]
```

Any suggestions?

**Comment by chaserileyroberts** [comment]
I think fundamentally my strategy about getting intermediary tensors by hand is flawed.

I'm going to redo this code with tfdbg, as they already have Nan and inf checks built-in.

See https://www.tensorflow.org/versions/master/programmers_guide/debugger#finding_nans_and_infs

I have to take a final right now, but I'll try to finish it later tonight

**Comment by dekromp** [comment]
Thx a lot. It is not super urgent so please go ahead and take care of your studies first.

**Comment by chaserileyroberts** [comment]
Alright finals are over and I actually got a few hours to work on this.

PR is merged, check it out here. https://github.com/Thenerdstation/mltest/pull/5

This issue should be fixed now. I ended up using tfdbg.DebugDumpDir to grab the intermediary values. It also now requires the latest versions of tensorflow, but that was just the version I use and can test easily. Let me know if this is a problem for your system and I'll try to lower it.
